### PR TITLE
feat(pr): remove and re-create automergeComment when rebasing

### LINKED ISSUE #5021: automergeComment should comment again after a rebase
**What would you like Renovate to be able to do?**
When using `automergeComment` and bors, we often get into a situation where a PR fails due to a merge conflict. Renovate will eventually rebase and resolve this merge conflict, but it will not issue another merge comment after this leaving us to merge them manually. 

**Describe the solution you'd like**
It would be great if `automergeComment` would comment again after a successful rebase. This could possibly be done if `ensureComment` took a timestamp of the last time a rebase happened and ensured there was a comment *after* that.

**Describe alternatives you've considered**
I've considered writing a bot to delete the comment so that it would get recreated. At that point, I may as well have a bot that auto accepts.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by ZauberNerd**
This PR will change the behavior of the `pr-comment` automerge strategy to try to remove and re-create the `automergeComment` after a rebase has hapened.
This is needed because some merge bots (such as `bors-ng`) do not try to merge again after a rebase has happened.

Update: This PR only addresses the original issue described in #5021 but does no longer address my proposed changes in https://github.com/renovatebot/renovate/issues/5021#issuecomment-584599546 because of some complications (https://github.com/renovatebot/renovate/pull/6002#pullrequestreview-396000479).

Closes #5021

### CODE REVIEW

File: lib/platform/common.ts [hunk 1]
```diff
@@ -3,6 +3,11 @@ import Git from 'simple-git/promise';
 import { RenovateConfig } from '../config/common';
 import { BranchStatus } from '../types';
 
+export interface Comment {
```
Comment (ZauberNerd) [inline lib/platform/common.ts:6]: We need to move the `Comment` interface here, because we now export the `getComments` method on the `Platform` interface: https://github.com/renovatebot/renovate/pull/6002/files#diff-660268298abca8fe83745a74ee3a6802R199

File: lib/platform/common.ts [hunk 2]
```diff
@@ -148,6 +153,7 @@ export interface EnsureCommentConfig {
   number: number;
   topic: string;
   content: string;
+  force?: boolean;
```
Comment (ZauberNerd) [inline lib/platform/common.ts:156]: The `force` option allows us to add a comment regardless of whether it exists already, which is necessary because the automerge pr-comment might not be authored by renovate and therefore cannot be removed by renovate.

File: lib/platform/github/index.ts [hunk 3]
```diff
@@ -1531,7 +1527,7 @@ async function deleteComment(commentId: number): Promise<void> {
   );
 }
 
-async function getComments(issueNo: number): Promise<Comment[]> {
+export async function getComments(issueNo: number): Promise<Comment[]> {
```
Comment (ZauberNerd) [inline lib/platform/github/index.ts:1530]: We need to export the `getComments` method because we want to find out if a automerge comment already exists: https://github.com/renovatebot/renovate/pull/6002/files#diff-5a59e8794d3cba4d86a62354df383482R435

**Review by ZauberNerd — COMMENTED** [review]
~~Hi @rarkins I would like some advise from you on how to continue with this PR.
This is the simplest solution to address the issue in #5021 but it does not remove previous comments (as you asked for: https://github.com/renovatebot/renovate/issues/5021#issuecomment-567419750).~~

~~I also had a look at `ensureCommentRemoval` (see here: https://github.com/renovatebot/renovate/compare/master...ZauberNerd:rebase-comment-v2?expand=1) but the problem with that is, that we'd still need to do the same thing as in this PR (exporting the `getComments` method and adding a `force` flag to the `ensureComment` method), because it might happen, that the automerge comment was not created by renovate and therefore cannot be removed.~~

~~Although, while I am thinking about this, I think I have found another problem, because now renovate would basically allow any GitHub member to write a `bors merge` comment on a PR and after a rebase renovate would repeat that comment allowing the PR to be merged, even if the author of the `bors merge` comment is not allowed to merge the PR.~~

~~I'd love to hear your opinions on this, meanwhile I'll also think about how to prevent arbitrary GitHub members to merge PRs on repositories where they wouldn't have rights to otherwise.~~

**Comment by viceice** [comment]
We should add a filter to getcomments function to only return renovate comments. So we only add or remove our own comments.

**Comment by ZauberNerd** [comment]
> We should add a filter to getcomments function to only return renovate comments. So we only add or remove our own comments.

I have removed the functionality that I requested in this comment: https://github.com/renovatebot/renovate/issues/5021#issuecomment-584599546, so now renovate will only try to remove and re-create automerge comments for PRs where automerge is enabled and therefore it is less likely that the PR already contains automerge comments that were not authored by renovate.

File: lib/platform/common.ts [hunk 4]
```diff
@@ -149,6 +149,11 @@ export interface EnsureCommentConfig {
   topic: string;
   content: string;
 }
+export interface EnsureCommentRemovalConfig {
```
Comment (rarkins) [inline lib/platform/common.ts:152]: Could you submit this refactoring as a separate PR that we merge first? Then the change in *functionality* in this PR will be more easily reviewed.

**Review by rarkins — CHANGES_REQUESTED** [review]
- Separate refactoring of ensureCommentRemoval signature into another PR
- Were you planning to implement this for all platforms once reviews for the GitHub implementation were done?

**Comment by ZauberNerd** [comment]
> * Separate refactoring of ensureCommentRemoval signature into another PR

Done: https://github.com/renovatebot/renovate/pull/6117
I have left the refactoring commit in here for now, but it should disappear, once the other PR has been merged. Or I could set the base branch of this PR to the branch of #6117 ?

> * Were you planning to implement this for all platforms once reviews for the GitHub implementation were done?

I was starting to, by just using the tests (because I don't have all of these SCMs at hand to verify), as the change should be fairly trivial, since it would basically be the same as removing a comment by topic. But when I started on azure, I found this:
https://github.com/renovatebot/renovate/blob/143ff5547a8fb8465b3bb4029466fcf801e60015/lib/platform/azure/index.spec.ts#L780-L795
Here the successful test checks that the azure API has been called 3 times.

Also, the test that checks that no comment is deleted if a comment could not be found, checks that the azure API has been called 3 times:
https://github.com/renovatebot/renovate/blob/143ff5547a8fb8465b3bb4029466fcf801e60015/lib/platform/azure/index.spec.ts#L801-L813

Judging by the code, the azure API should only be called 2 times when a comment has not been found:
https://github.com/renovatebot/renovate/blob/143ff5547a8fb8465b3bb4029466fcf801e60015/lib/platform/azure/index.ts#L576-L603

So after I found that, I concentrated on the GitHub API and left the others blank for now, because I wasn't sure how to proceed

Should all other SCMs be implemented as well?

**Comment by viceice** [comment]
I think we should support it for all platforms. Maybe we should first add filtering my comments, so renovate will only ever see his own comments.
